### PR TITLE
`SignatureScheme` trait bounds

### PR DESCRIPTION
**Pull request by tessico**
## Description

Add trait bounds on `SignatureScheme` trait.
Bounds not yet added for `SigningKey`, pending discussion in [espresso-common](https://github.com/EspressoSystems/espresso-systems-common/issues/7).

closes: #145, #155

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (main)
- [x] Linked to GitHub issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Wrote unit tests
- [x] Updated relevant documentation in the code
- [ ] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the GitHub PR explorer

### CODE REVIEW

**Comment by tessico** [comment]
@EspressoSystems/jellyfish

File: primitives/src/signatures/bls.rs [hunk 1]
```diff
@@ -3,16 +3,136 @@
...
-    PublicKey as BLSVerKey, SecretKey as BLSSignKey, Signature as BLSSignature,
-};
+pub use blst::min_sig::{PublicKey, SecretKey, Signature};
+
+/// Size in bytes of a secret key in our BLS signature scheme.
+pub const BLS_SIG_KEY_SIZE: usize = 32;
+/// Size in bytes of a signature in our BLS signature scheme.
+pub const BLS_SIG_SIGNATURE_SIZE: usize = 96;
+/// Size in bytes of a verification key in our BLS signature scheme.
+pub const BLS_SIG_VERKEY_SIZE: usize = 192;
```
[the remark this answers is not in the document]
Comment (tessico) [reply primitives/src/signatures/bls.rs]: Done in https://github.com/EspressoSystems/jellyfish/pull/148/commits/f14cca08df814306663f9d8f5c65d8fc96be02dd.

File: primitives/src/signatures/bls.rs [hunk 2]
```diff
@@ -3,16 +3,136 @@
...
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct BLSSignature(Signature);
+
+impl core::ops::Deref for BLSSignature {
+    type Target = Signature;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
```
[the remark this answers is not in the document]
Comment (alxiong) [reply primitives/src/signatures/bls.rs:72]: yeah like the anti-pattern page says: it's abusing the `Deref` which is supposed to be custom pointer type, not newtype accessing its inner value. (another testimonyhttps://users.rust-lang.org/t/how-to-properly-use-newtypes-to-extend-a-struct/34042 of the same problem)

I don't mind too much about `sig.0.verify()`, it's internal details, won't get exposed to consumer anyway.

File: primitives/src/signatures/mod.rs [hunk 3]
```diff
@@ -17,16 +18,16 @@ pub trait SignatureScheme {
     const CS_ID: &'static str;
 
     /// Signing key.
-    type SigningKey;
+    type SigningKey: Clone + Send + Sync;
```
[the remark this answers is not in the document]
Comment (tessico) [reply primitives/src/signatures/mod.rs]: Sure! Schnorr signature already derives it, so it's minimal change effort. Done in https://github.com/EspressoSystems/jellyfish/pull/148/commits/202148e47522b1fa4443e7ba4e7781b7607f10c4.

File: primitives/src/signatures/bls.rs [hunk 4]
```diff
@@ -3,16 +3,136 @@
...
+impl CanonicalDeserialize for BLSSignKey {
+    fn deserialize<R: ark_serialize::Read>(mut reader: R) -> Result<Self, SerializationError> {
+        let len = <usize as ark_serialize::CanonicalDeserialize>::deserialize(&mut reader)?;
+        if len != BLS_SIG_KEY_SIZE {
+            return Err(SerializationError::InvalidData);
+        }
+
+        let mut key = [0u8; BLS_SIG_KEY_SIZE];
+        reader.read_exact(&mut key)?;
+        Ok(Self(SecretKey::deserialize(&key).unwrap()))
```
[the remark this answers is not in the document]
Comment (tessico) [reply primitives/src/signatures/bls.rs]: Done for all new structs in https://github.com/EspressoSystems/jellyfish/pull/148/commits/a00c7691f4c3e15610b0ef42920c1f98d9471df3.

**Review by alxiong — DISMISSED** [review]
LGTM!

File: primitives/src/signatures/bls.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (alxiong) [reply primitives/src/signatures/bls.rs]: So as I dig deeper, there are some subtleties involved.

TL;DR: **yes, in our case, we can use `Deref` as you originally proposed.** @tessico 

this comment https://github.com/rust-lang/api-guidelines/issues/249#issuecomment-904446836 convinced me that using `Deref` for transparent newtype is not only okay but also practically seen in std lib and some high-profile crates.
So suggested [in the Rust book itself](https://doc.rust-lang.org/book/ch19-03-advanced-traits.html?highlight=Deref#using-the-newtype-pattern-to-implement-external-traits-on-external-types).

What I'm also more sure of is:
- we should NOT use `Deref` to simulate OOP-style inheritance (instead, we should wait for delegation to be available in rust, or use [`ambassador` crate](https://crates.io/crates/ambassador) to achieve that)
- we should be careful about the potential violation of intended visibility when implementing `DerefMut` (to access `fn method(&mut self)` of the underlying types).  see my [example here](https://twitter.com/alex_xiong_/status/1598184911986315264).
- Jon had updated a good [summary here](https://github.com/jonhoo/rust-for-rustaceans.com/commit/c59c732488f9322d25013512acc19d1bba111c6e).

[the remark this answers is not in the document]
Comment (tessico) [reply primitives/src/signatures/bls.rs]: @alxiong Nice research! I knew I saw it in the Rust book somewhere, I must have missed it last time around.
It's encouraging to see it used in std :)

File: primitives/src/signatures/mod.rs [hunk 5]
```diff
@@ -17,16 +19,16 @@ pub trait SignatureScheme {
...
 
     /// Public Parameter
     type PublicParameter;
 
     /// Signature
-    type Signature;
+    type Signature: Clone + for<'a> Deserialize<'a> + Serialize;
 
     /// A message is &\[MessageUnit\]
     type MessageUnit;
```
[the remark this answers is not in the document]
Comment (tessico) [reply primitives/src/signatures/mod.rs]: It seems that `blst::min_sig::SecretKey` doesn't implement `PartialEq` for some reason (while `PublicKey` and `Signature` do), see an issue [here](https://github.com/supranational/blst/issues/12). For now I've implemented it manually by comparing the output of serialization, which seems the only way currently.
Depending on what ppl say in the blst issue, we might either add the implementation to their crate and contribute upstream, or if there was a good reason for not including it, remove the bounds from our `SignatureScheme` trait.

[the remark this answers is not in the document]
Comment (alxiong) [reply primitives/src/signatures/mod.rs]: oh we have overlapping work. I also implemented a (possibly constant-time) comparison for `PartialEq` in #156

**Review by alxiong — APPROVED** [review]
LGTM! 

(I will try to resolve the conflict with my PR later, somehow some of the overlap between our work slip my mind when I worked on my PR)
